File: LLD/Design-Patterns/Adapter/code/Adapter.cpp
// Adapter Pattern — adapting an incompatible third-party payment SDK.
// Compile: g++ -std=c++14 Adapter.cpp -o adapter
// Run:     ./adapter

#include <iostream>
#include <string>
#include <cmath>

class PaymentProcessor {
public:
    virtual void pay(double amountDollars) = 0;
    virtual ~PaymentProcessor() = default;
};

class InHousePaymentProcessor : public PaymentProcessor {
public:
    void pay(double amountDollars) override {
        std::cout << "In-house processor charged $" << amountDollars << std::endl;
    }
};

// The third-party SDK we don't control — incompatible method name and units.
class ThirdPartySDK {
public:
    void makeTransaction(int amountCents, const std::string& currencyCode) {
        std::cout << "ThirdPartySDK processed " << amountCents << " " << currencyCode << " cents" << std::endl;
    }
};

class ThirdPartyPaymentAdapter : public PaymentProcessor {
public:
    explicit ThirdPartyPaymentAdapter(ThirdPartySDK& sdk) : sdk_(sdk) {}

    void pay(double amountDollars) override {
        int cents = static_cast<int>(std::round(amountDollars * 100));
        sdk_.makeTransaction(cents, "USD");
    }

private:
    ThirdPartySDK& sdk_;
};

void checkout(PaymentProcessor& processor, double amount) {
    processor.pay(amount);
}

int main() {
    InHousePaymentProcessor inHouse;
    ThirdPartySDK sdk;
    ThirdPartyPaymentAdapter thirdParty(sdk);

    checkout(inHouse, 49.99);
    checkout(thirdParty, 49.99);

    return 0;
}

File: v1/LLD/Design-Patterns/Adapter/code/adapter.py
"""
Adapter Pattern — adapting an incompatible third-party payment SDK.
Run: python adapter.py
"""

from abc import ABC, abstractmethod


class PaymentProcessor(ABC):
    @abstractmethod
    def pay(self, amount_dollars):
        ...


class InHousePaymentProcessor(PaymentProcessor):
    def pay(self, amount_dollars):
        print(f"In-house processor charged ${amount_dollars}")


class ThirdPartySDK:
    def make_transaction(self, amount_cents, currency_code):
        print(f"ThirdPartySDK processed {amount_cents} {currency_code} cents")


class ThirdPartyPaymentAdapter(PaymentProcessor):
    def __init__(self, sdk: ThirdPartySDK):
        self._sdk = sdk

    def pay(self, amount_dollars):
        cents = round(amount_dollars * 100)
        self._sdk.make_transaction(cents, "USD")


def checkout(processor: PaymentProcessor, amount):
    processor.pay(amount)


if __name__ == "__main__":
    in_house = InHousePaymentProcessor()
    third_party = ThirdPartyPaymentAdapter(ThirdPartySDK())

    checkout(in_house, 49.99)
    checkout(third_party, 49.99)
